### PR TITLE
eslint: require fix for suggestions

### PR DESCRIPTION
**Pull request by 43081j**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).
- [ ] Provide a URL to documentation or source code which provides context for the suggested changes: https://eslint.org/docs/developer-guide/working-with-rules#providing-suggestions

---

When providing a suggestion in eslint, a fixer _must_ exist. It cannot be `undefined` or `null`, so our current types are incorrect by reusing the report base.

I could've probably used something like `Required<T>` or similar to re-use the base but i figured its more readable just having two interfaces.

I've also removed a redundant `undefined` (surprisingly the `no-redundant-undefined` rule hasn't been complaining about this).

### CODE REVIEW

**Review by bradzacher — APPROVED** [review]

File: types/eslint/index.d.ts [hunk 1]
```diff
@@ -597,14 +597,22 @@ export namespace Rule {
         report(descriptor: ReportDescriptor): void;
     }
 
+    type ReportFixer = ((fixer: RuleFixer) => null | Fix | IterableIterator<Fix> | Fix[]);
```
Comment (JounQin) [inline types/eslint/index.d.ts:600]: Is the wrapping `()` needed here?

Comment (43081j) [reply types/eslint/index.d.ts:600]: it isn't, i had copied it from the previous code. will update, good catch 👍

**Review by JounQin — APPROVED** [review]

**Review by gabritto — APPROVED** [review]
Looks good to me. The "redundant" `undefined`s you removed are not redundant anymore, starting from TypeScript 4.3, if you have [`exactOptionalPropertyTypes`](https://www.typescriptlang.org/tsconfig#exactOptionalPropertyTypes) option set. But it's still quite some time until we turn that on for DT.

**Comment by 43081j** [comment]
Ready to merge 🥳

**Pull request merged**